### ISSUE TITLE
An error 'Disallowed method "elementsFromPoint" on ShadowRoot.' is thrown when I use Selenium moveToElement() to Highcharts' element

### ISSUE THREAD

**Issue by nobyk** [salesforce/lwc#2333]
### Description
We have a Selenium test which checks if hover action works on Highcharts' bars. It worked fine on a scratch org with Spring '21 Patch 14.4. But on a scratch org with Spring '21 Patch 16.1 or later, it started getting an error 'Disallowed method "elementsFromPoint" on ShadowRoot.'.

#### Steps to Reproduce
1. Create a Highchart which has a tooltip at mouseover.
2. Run a Selenium test which calls moveToElement() to a bar element `rect` in the chart.

#### Expected Results
No error is thrown and a tooltip is shown.

#### Actual Results
An error 'Disallowed method "elementsFromPoint" on ShadowRoot.' is thrown

### Browsers Affected
Chrome WebDriver

### Version
- Spring '21 Patch 16.1 (Sorry, I'm not sure the LWC version)

**Comment by ekashida** [comment]
This exception has been in place for the past several years:
https://github.com/salesforce/lwc/blob/fb727400ffb4c1a8938169ff06786ff525279981/packages/lwc-engine/src/faux-shadow/shadow-root.ts#L235-L238

We recently improved the messaging by adding a bit more detail:
https://github.com/salesforce/lwc/pull/2192/files#diff-550981a0270159a5ce9f55112b56a6c86e0aa7ffc285fd4f6caf951ef134b62f

Is it possible that the exception was less visible before due to the lack of an actual error message?

**Comment by nobyk** [comment]
> Is it possible that the exception was less visible before due to the lack of an actual error message?
It seems not possible. Selenium gets an exception if the error is thrown at the point and the test is marked as failed, so it can hardly be missed.

**Comment by pmdartus** [comment]
Could you share a small reproduction code for us to try out? As @ekashida except for the error message, this code hasn't changed for a while now.

**Comment by nobyk** [comment]
Here is a sample C# code to reproduce the error.

```
var rect = webDriver.FindElement(By.XPath("XPath to your bar rect in a highchart"));
new Actions(webDriver).MoveToElement(rect).Perform();
```

**Comment by nolanlawson** [comment]
As I understand it, [`MoveToElement()`](https://www.selenium.dev/selenium/docs/api/java/org/openqa/selenium/interactions/Actions.html#moveToElement(org.openqa.selenium.WebElement)) causes a hover event to occur. Could it be that something changed in Selenium/Highchart such that it is calling `elementsFromPoint()` now when it didn't before?

Also, can you reproduce manually (without Selenium/WebDriver) by just hovering the chart element? Does this also throw an error?

I tried to reproduce your error manually with just LWC + Synthetic Shadow + Highcharts, but was unable. [Here is my code](https://github.com/nolanlawson/lwc-barebone/commit/88134c4d4db59702008522757235409efc391406) (you can run it by checking out the code and running `yarn && yarn build && yarn start`).

**Comment by nobyk** [comment]
> Could it be that something changed in Selenium/Highchart such that it is calling elementsFromPoint() now when it didn't before?

I think there is no change in Selenium because the error doesn't occur in Spring '21 Patch 14.4 with same version of Selenium.

> Also, can you reproduce manually (without Selenium/WebDriver) by just hovering the chart element? Does this also throw an error?

It cannot be manually reproduced.

**Comment by nolanlawson** [comment]
I can't seem to reproduce using my sample app and Selenium + WebDriverIO.

For whatever reason, it seems that hovering is triggering `elementsFromPoint` on the shadowRoot, which is supposed to throw an error (working as intended). Is it possible to use a focus event instead of a hover event to get the tooltip to appear?

<details>
<summary>WebDriverIO code</summary>

Run the sample LWC app from above. Then run `npm i -g selenium-standalone && selenium-standalone install & &selenium-standalone start` in another terminal.


Then install webdriverio `npm install webdriverio` and run:

```js
const { remote } = require('webdriverio')

const main = async () => {
  const browser = await remote({
    capabilities: {
      browserName: 'chrome'
    },
    hostname: 'localhost',
    port: 4444,
    automationProtocol: 'webdriver',
    path: '/wd/hub'
  })

  await browser.url('http://localhost:5000')
  const element = await browser.$(() => document.querySelector('x-app').shadowRoot.querySelector('rect'))
  await element.moveTo()
  await browser.closeWindow()
}

main()

```
</details>

**Comment by nobyk** [comment]
> I can't seem to reproduce using my sample app and Selenium + WebDriverIO.

Could you try to add following `tooptip` configuration to your chart?
```
tooltip: {
    useHTML: true,
    animation: false,
    formatter: function () {
        return 'any html for your tooltip';
    }
}
```

> For whatever reason, it seems that hovering is triggering `elementsFromPoint` on the shadowRoot, which is supposed to throw an error (working as intended).

Is it possible that lwc wrongly detects `elementsFromPoint` on the shadowRoot? I wonder why it started getting the error recent version of lwc.

**Comment by nobyk** [comment]
There is a post querying same issue on Stack Overflow.
https://stackoverflow.com/questions/67483795/javascript-error-disallowed-method-elementsfrompoint-on-shadowroot-on-doing-m

They do a workaround focusing adjoining element and then send TAB key. But we cannot use the trick for columns in the chart.

**Comment by nolanlawson** [comment]
I'm sorry, but I still cannot reproduce. I tried [adding the tooltip](https://github.com/nolanlawson/lwc-barebone/commit/d80d09d5933a5179122a2d6557185fd4ce6f5f06) and focusing the `<rect>`, but the error doesn't repro.

I think ultimately though the only solution may be the one suggested in the StackOverflow post – using focus/tab events instead. It seems that something about `moveTo*` is causing Selenium/WebDriver to try to call `shadowRoot.elementsFromPoint`, which is disallowed as mentioned above.

**Comment by nobyk** [comment]
But the error was implemented on 21 Oct 2018, right? Why it started occurring on recent version of lwc? There is no change on Selenium side.

**Comment by nobyk** [comment]
I checked our test execution log. Our ChromeDriver is updated to 90.0.4430.2400 from 89.0.4389.2300 when it started getting the error. So `elementsFromPoint` behavior could be changed on the version. I'll close the issue.

Thank you all for your kind cooperation.

**Comment by nobyk** [comment]
I could confirm that the error has been mitigated with ChromeDriver 89.0.4389.2300. Let me thank you again.

**Comment by nolanlawson** [comment]
Thanks for the details. I still can't seem to find a repro, but the change the caused the issue was probably this one from January of 2020: https://github.com/bayandin/chromedriver/commit/ad6ede81b0d048e031fbe7cc877dcb7d7b1f0049

As you can see, it's adding a call to `element.getRootNode().elementsFromPoint()`, which would definitely throw an error in synthetic shadow DOM.

**Comment by nobyk** [comment]
@nolanlawson,
Thank you for your check. I use JavaScript to trigger `mouseover` event instead and it works.

```
var rect = webDriver.FindElement(By.XPath("XPath to your bar rect in a highchart"));
var javaScriptExecutor = (IJavaScriptExecutor)webDriver;
javaScriptExecutor.ExecuteScript("arguments[0].dispatchEvent(new MouseEvent('mouseover', {'bubbles': true }));", rect);
```

**Comment by nolanlawson** [comment]
There is now [a Knowledge Article](https://help.salesforce.com/articleView?id=000358800&type=1&mode=1) about this.

**Comment by bereg2k** [comment]
@nolanlawson Hi! When can we expect #2381 fix to go live?

**Comment by nolanlawson** [comment]
@bereg2k It was released in LWC [v2.2.6](https://github.com/salesforce/lwc/releases/tag/v2.2.6). If you're asking about Salesforce platform releases, it looks like it made it into Winter '22. You can check [status.salesforce.com](https://status.salesforce.com/) to see when this release will reach your instance. 🙂

**Comment by bereg2k** [comment]
@nolanlawson yes, platform release indeed. Thanks for the answer! Looking forward to this fix. :slightly_smiling_face:
